compiler/src/pipeline.rs

```diff
@@ -0,0 +1,489 @@
+///      vec![],
+/// )
+/// .unwrap();
+/// pipeline.prove(BackendType::PilStarkCli).unwrap();
+/// ```
+impl<T: FieldElement> Pipeline<T> {
+    pub fn with_tmp_output(self) -> Self {
+        let tmp_dir = mktemp::Temp::new_dir().unwrap();
+        Pipeline {
+            output_dir: Some(tmp_dir.to_path_buf()),
+            tmp_dir: Some(tmp_dir),
+            ..self
+        }
+    }
+
+    pub fn with_output(self, output_dir: PathBuf, force_overwrite: bool) -> Self {
+        Pipeline {
+            output_dir: Some(output_dir),
+            force_overwrite,
+            ..self
+        }
+    }
+
+    pub fn from_file(self, asm_file: PathBuf) -> Self {
+        if asm_file.extension().unwrap() == "asm" {
+            self.from_asm_file(asm_file)
+        } else {
+            self.from_pil_file(asm_file)
+        }
+    }
+
+    pub fn from_asm_file(self, asm_file: PathBuf) -> Self {
+        let name = self.name.or(Some(Self::name_from_path(&asm_file)));
+        Pipeline {
+            artifact: Some(Artifact::AsmFile(asm_file)),
+            name,
+            ..self
+        }
+    }
+
+    pub fn from_asm_string(self, asm_string: String, path: Option<PathBuf>) -> Self {
+        let name = self.name.or(path.as_ref().map(|p| Self::name_from_path(p)));
+        Pipeline {
+            artifact: Some(Artifact::AsmString(path, asm_string)),
+            name,
+            ..self
+        }
+    }
+
+    pub fn from_pil_file(self, pil_file: PathBuf) -> Self {
+        let name = self.name.or(Some(Self::name_from_path(&pil_file)));
+        Pipeline {
+            artifact: Some(Artifact::PilFile(pil_file)),
+            name,
+            ..self
+        }
+    }
+
+    pub fn from_pil_string(self, pil_string: String) -> Self {
+        Pipeline {
+            artifact: Some(Artifact::PilString(pil_string)),
+            ..self
+        }
+    }
+
+    fn name_from_path(path: &Path) -> String {
+        path.file_stem().unwrap().to_str().unwrap().to_string()
+    }
+
+    fn log(&self, msg: &str) {
+        log::log!(self.log_level, "{}", msg);
+    }
+
+    #[allow(clippy::print_stderr)]
+    fn advance(&mut self) -> Result<(), Vec<String>> {
+        let artifact = std::mem::take(&mut self.artifact).unwrap();
+        self.artifact = Some(match artifact {
+            Artifact::AsmFile(path) => Artifact::AsmString(
+                Some(path.clone()),
+                fs::read_to_string(&path).map_err(|e| {
+                    vec![format!("Error reading .asm file: {}\n{e}", path.display())]
+                })?,
+            ),
+            Artifact::AsmString(path, asm_string) => {
+                let parsed_asm = parser::parse_asm(None, &asm_string).unwrap_or_else(|err| {
+                    match path.as_ref() {
+                        Some(path) => eprintln!("Error parsing .asm file: {}", path.display()),
+                        None => eprintln!("Error parsing .asm file:"),
```

Comment:
Why not return `Err(error string from the ones above)`?

Reply:
Hmm, copied this, not sure if there is a good reason. The error you get here is a `ParseError` which only implements `output_to_stderr`. Maybe it's about support for colored output or something like this?

Reply:
ok let's keep